README.md
# ulid

[![Build Status](https://travis-ci.org/ahawker/ulid.svg?branch=master)](https://travis-ci.org/ahawker/ulid)
[![Build Status](https://ci.appveyor.com/api/projects/status/fy0hufnb8h6gwk4d/branch/master?svg=true)](https://ci.appveyor.com/project/ahawker/ulid/branch/master)
[![codecov](https://codecov.io/gh/ahawker/ulid/branch/master/graph/badge.svg)](https://codecov.io/gh/ahawker/ulid)
[![Code Climate](https://codeclimate.com/github/ahawker/ulid/badges/gpa.svg)](https://codeclimate.com/github/ahawker/ulid)
[![Issue Count](https://codeclimate.com/github/ahawker/ulid/badges/issue_count.svg)](https://codeclimate.com/github/ahawker/ulid)

[![PyPI Version](https://badge.fury.io/py/ulid-py.svg)](https://badge.fury.io/py/ulid-py)
[![PyPI Versions](https://img.shields.io/pypi/pyversions/ulid-py.svg)](https://pypi.python.org/pypi/ulid-py)

[![Updates](https://pyup.io/repos/github/ahawker/ulid/shield.svg)](https://pyup.io/repos/github/ahawker/ulid/)

[![Documentation Status](https://readthedocs.org/projects/ulid/badge/?version=latest)](http://ulid.readthedocs.io/en/latest/?badge=latest)

[Universally Unique Lexicographically Sortable Identifier](https://github.com/alizain/ulid) in [Python 3](https://www.python.org/).

### Status

This project is actively maintained.

### Installation

To install ulid from [pip](https://pypi.python.org/pypi/pip):
```bash
    $ pip install ulid-py
```

To install ulid from source:
```bash
    $ git clone git@github.com:ahawker/ulid.git
    $ cd ulid && python setup.py install
```

### Usage

Create a brand new ULID.

The timestamp value (48-bits) is from [time.time()](https://docs.python.org/3/library/time.html?highlight=time.time#time.time) with millisecond precision.

The randomness value (80-bits) is from [os.urandom()](https://docs.python.org/3/library/os.html?highlight=os.urandom#os.urandom).

```python
>>> import ulid
>>> ulid.new()
<ULID('01BJQE4QTHMFP0S5J153XCFSP9')>
```

Create a new ULID from an existing 128-bit value, such as a [UUID](https://docs.python.org/3/library/uuid.html).

Supports ULID values as `int`, `bytes`, `str`, and `UUID` types.

```python
>>> import ulid, uuid
>>> value = uuid.uuid4()
>>> value
UUID('0983d0a2-ff15-4d83-8f37-7dd945b5aa39')
>>> ulid.from_uuid(value)
<ULID('09GF8A5ZRN9P1RYDVXV52VBAHS')>
```

Create a new ULID from an existing timestamp value, such as a [datetime](https://docs.python.org/3/library/datetime.html#module-datetime) object.

Supports timestamp values as `int`, `float`, `str`, `bytes`, `bytearray`, `memoryview`, `datetime`, `Timestamp`, and `ULID` types.

```python
>>> import datetime, ulid
>>> ulid.from_timestamp(datetime.datetime(1999, 1, 1))
<ULID('00TM9HX0008S220A3PWSFVNFEH')>
```

Create a new ULID from an existing randomness value.

Supports randomness values as `int`, `float`, `str`, `bytes`, `bytearray`, `memoryview`, `Randomness`, and `ULID` types.

```python
>>> import os, ulid
>>> randomness = os.urandom(10)
>>> ulid.from_randomness(randomness)
>>> <ULID('01BJQHX2XEDK0VN0GMYWT9JN8S')>
```

For cases when you don't necessarily control the data type (input from external system), you can use the `parse` method
which will attempt to make the correct determination for you. Please note that this will be slightly slower than creating
the instance from the respective `from_*` method as it needs to make a number of type/conditional checks.

Supports values as `int`, `float`, `str`, `bytes`, `bytearray`, `memoryview`, `uuid.UUID`, and `ULID` types.

```python
>>> import ulid
>>> value = db.model.get_id()  ## Unsure about datatype -- Could be int, UUID, or string?
>>> ulid.parse(value)
>>> <ULID('0K0EDFETFM8SH912DBBD4ABXSZ')>
```

Once you have a ULID object, there are a number of ways to interact with it.

The `timestamp` method will give you a snapshot view of the first 48-bits of the ULID while the `randomness` method
will give you a snapshot of the last 80-bits.

```python
>>> import ulid
>>> u = ulid.new()
>>> u
<ULID('01BJQM7SC7D5VVTG3J68ABFQ3N')>
>>> u.timestamp()
<Timestamp('01BJQM7SC7')>
>>> u.randomness()
<Randomness('D5VVTG3J68ABFQ3N')>
```

The `ULID`, `Timestamp`, and `Randomness` classes all derive from the same base class, a `MemoryView`.

A `MemoryView` provides the `bin`, `bytes`, `hex`, `int`, `oct`, and `str`, methods for changing any values representation.

```python
>>> import ulid
>>> u = ulid.new()
>>> u
<ULID('01BJQMF54D093DXEAWZ6JYRPAQ')>
>>> u.timestamp()
<Timestamp('01BJQMF54D')>
>>> u.timestamp().int
1497589322893
>>> u.timestamp().bytes
b'\x01\\\xafG\x94\x8d'
>>> u.timestamp().datetime
datetime.datetime(2017, 6, 16, 5, 2, 2, 893000)
>>> u.randomness().bytes
b'\x02F\xde\xb9\\\xf9\xa5\xecYW'
>>> u.bytes[6:] == u.randomness().bytes
True
>>> u.str
'01BJQMF54D093DXEAWZ6JYRPAQ'
>>> u.int
1810474399624548315999517391436142935
>>> u.bin
'0b1010111001010111101000111100101001000110100000010010001101101111010111001010111001111100110100101111011000101100101010111'
>>> u.hex
'0x15caf47948d0246deb95cf9a5ec5957'
>>> u.oct
'0o12712750745106402215572712717464573054527'
```

A `MemoryView` also provides rich comparison functionality.

```python
>>> import datetime, time, ulid
>>> u1 = ulid.new()
>>> time.sleep(5)
>>> u2 = ulid.new()
>>> u1 < u2
True
>>> u3 = ulid.from_timestamp(datetime.datetime(2039, 1, 1))
>>> u1 < u2 < u3
True
>>> [u.timestamp().datetime for u in sorted([u2, u3, u1])]
[datetime.datetime(2017, 6, 16, 5, 7, 14, 847000), datetime.datetime(2017, 6, 16, 5, 7, 26, 775000), datetime.datetime(2039, 1, 1, 8, 0)]
```

### Future Items

* Collection of benchmarks to track performance.
* Backport to Python 2.7?
* See [Github Issues](https://github.com/ahawker/ulid/issues) for more!

### Goals

A fast implementation in pure python of the spec with binary format support.

### Contributing

If you would like to contribute, simply fork the repository, push your changes and send a pull request.
Pull requests will be brought into the `master` branch via a rebase and fast-forward merge with the goal of having a linear branch history with no merge commits.

### License

[Apache 2.0](LICENSE)

## Why not UUID?

UUID can be suboptimal for many uses-cases because:

* It isn't the most character efficient way of encoding 128 bits of randomness
* UUID v1/v2 is impractical in many environments, as it requires access to a unique, stable MAC address
* UUID v3/v5 requires a unique seed and produces randomly distributed IDs, which can cause fragmentation in many data structures
* UUID v4 provides no other information than randomness which can cause fragmentation in many data structures

ULID provides:

* 128-bit compatibility with UUID
* 1.21e+24 unique ULIDs per millisecond
* Lexicographically sortable!
* Canonically encoded as a 26 character string, as opposed to the 36 character UUID
* Uses Crockford's base32 for better efficiency and readability (5 bits per character)
* Case insensitive
* No special characters (URL safe)

## Specification

Below is the current specification of ULID as implemented in this repository.

The binary format is implemented.

```
 01AN4Z07BY      79KA1307SR9X4MV3

|----------|    |----------------|
 Timestamp          Randomness
  10chars            16chars
   48bits             80bits
```

### Components

**Timestamp**
* 48 bit integer
* UNIX-time in milliseconds
* Won't run out of space till the year 10895 AD.

**Randomness**
* 80 bits
* Cryptographically secure source of randomness, if possible

### Sorting

The left-most character must be sorted first, and the right-most character sorted last (lexical order).
The default ASCII character set must be used. Within the same millisecond, sort order is not guaranteed

### Encoding

Crockford's Base32 is used as shown. This alphabet excludes the letters I, L, O, and U to avoid confusion and abuse.

```
0123456789ABCDEFGHJKMNPQRSTVWXYZ
```

### Binary Layout and Byte Order

The components are encoded as 16 octets. Each component is encoded with the Most Significant Byte first (network byte order).

```
0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                      32_bit_uint_time_high                    |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|     16_bit_uint_time_low      |       16_bit_uint_random      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                       32_bit_uint_random                      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                       32_bit_uint_random                      |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

### String Representation

```
ttttttttttrrrrrrrrrrrrrrrr

where
t is Timestamp
r is Randomness
```

### Links

* [Original Implementation (Javascript)](https://github.com/alizain/ulid)
* [ulid (python)](https://github.com/mdipierro/ulid)


CHANGELOG.md
# Changelog

## [v0.2.0](https://github.com/ahawker/ulid/tree/v0.2.0) (2020-08-11)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.1.0...v0.2.0)

**Implemented enhancements:**

- Values for range queries [\#457](https://github.com/ahawker/ulid/issues/457)

**Closed issues:**

- Add bounds checking for max timestamp overflow case [\#468](https://github.com/ahawker/ulid/issues/468)

**Merged pull requests:**

- Add MSB validation to raise when timestamp overflows 128 bits [\#471](https://github.com/ahawker/ulid/pull/471) ([ahawker](https://github.com/ahawker))
- Freeze isort to 4.x on py36 [\#470](https://github.com/ahawker/ulid/pull/470) ([ahawker](https://github.com/ahawker))
- Add API create method and min/max constants [\#469](https://github.com/ahawker/ulid/pull/469) ([ahawker](https://github.com/ahawker))
- Add support for bin, oct, and hex representations [\#466](https://github.com/ahawker/ulid/pull/466) ([ahawker](https://github.com/ahawker))
- pyup.io:  Scheduled weekly dependency update for week 31 [\#464](https://github.com/ahawker/ulid/pull/464) ([pyup-bot](https://github.com/pyup-bot))
- pyup.io:  Scheduled weekly dependency update for week 25 [\#456](https://github.com/ahawker/ulid/pull/456) ([pyup-bot](https://github.com/pyup-bot))

## [v0.1.0](https://github.com/ahawker/ulid/tree/v0.1.0) (2020-06-17)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.14...v0.1.0)

**Closed issues:**

- deepcopy doesn't work on ULID object [\#452](https://github.com/ahawker/ulid/issues/452)
- Problems with using ulid with `mypy --strict` [\#450](https://github.com/ahawker/ulid/issues/450)

**Merged pull requests:**

- Add pickle support [\#455](https://github.com/ahawker/ulid/pull/455) ([ahawker](https://github.com/ahawker))
- Explicitly define public api in \_\_init\_\_.py [\#454](https://github.com/ahawker/ulid/pull/454) ([ahawker](https://github.com/ahawker))
- Support mypy strict [\#453](https://github.com/ahawker/ulid/pull/453) ([ahawker](https://github.com/ahawker))
- pyup.io:  Scheduled weekly dependency update for week 24 [\#451](https://github.com/ahawker/ulid/pull/451) ([pyup-bot](https://github.com/pyup-bot))
- pyup.io:  Scheduled weekly dependency update for week 16 [\#442](https://github.com/ahawker/ulid/pull/442) ([pyup-bot](https://github.com/pyup-bot))

## [v0.0.14](https://github.com/ahawker/ulid/tree/v0.0.14) (2020-04-25)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.13...v0.0.14)

**Merged pull requests:**

- fix py.typed path [\#441](https://github.com/ahawker/ulid/pull/441) ([kentac55](https://github.com/kentac55))
- pyup.io:  Scheduled weekly dependency update for week 06 [\#432](https://github.com/ahawker/ulid/pull/432) ([pyup-bot](https://github.com/pyup-bot))
- pyup.io:  Scheduled weekly dependency update for week 05 [\#431](https://github.com/ahawker/ulid/pull/431) ([pyup-bot](https://github.com/pyup-bot))

## [v0.0.13](https://github.com/ahawker/ulid/tree/v0.0.13) (2020-04-25)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.12...v0.0.13)

## [v0.0.12](https://github.com/ahawker/ulid/tree/v0.0.12) (2019-11-22)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.11...v0.0.12)

**Merged pull requests:**

- Add \_\_bytes\_\_ support to MemoryView [\#430](https://github.com/ahawker/ulid/pull/430) ([ahawker](https://github.com/ahawker))
- Add float support to MemoryView [\#429](https://github.com/ahawker/ulid/pull/429) ([ahawker](https://github.com/ahawker))

## [v0.0.11](https://github.com/ahawker/ulid/tree/v0.0.11) (2019-11-20)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.10...v0.0.11)

**Merged pull requests:**

- Change travis deploy version to 3.7 so there aren't duplicates [\#428](https://github.com/ahawker/ulid/pull/428) ([ahawker](https://github.com/ahawker))
- Remove README.rst and switch setup.py to use markdown version [\#427](https://github.com/ahawker/ulid/pull/427) ([ahawker](https://github.com/ahawker))

## [v0.0.10](https://github.com/ahawker/ulid/tree/v0.0.10) (2019-11-20)

[Full Changelog](https://github.com/ahawker/ulid/compare/v0.0.9...v0.0.10)

**Implemented enhancements:**

- Backport to Python 2.7? [\#1](https://github.com/ahawker/ulid/issues/1)

**Merged pull requests:**

- Add PEP561 Support [\#426](https://github.com/ahawker/ulid/pull/426) ([ahawker](https://github.com/ahawker))
- pyup.io:  Scheduled weekly dependency update for week 44 [\#425](https://github.com/ahawker/ulid/pull/425) ([pyup-bot](https://github.com/pyup-bot))
- pyup.io:  Scheduled weekly dependency update for week 43 [\#424](https://github.com/ahawker/ulid/pull/424) ([pyup-bot](https://github.com/pyup-bot))
- Add python 3.8 support [\#423](https://github.com/ahawker/ulid/pull/423) ([ahawker](https://github.com/ahawker))
- pyup.io:  Scheduled weekly dependency update for week 42 [\#422](https://github.com/ahawker/ulid/pull/422) ([pyup-bot](https://github.com/pyup-bot))
- pyup.io:  Scheduled daily dependency update on Thursday [\#420](https://github.com/ahawker/ulid/pull/420) ([pyup-bot](https://github.com/pyup-bot))
- Update CHANGELOG.md based on latest [\#417](https://github.com/ahawker/ulid/pull/417) ([ahawker](https://github.com/ahawker))
- Update README.rst from latest README.md [\#416](https://github.com/ahawker/ulid/pull/416) ([ahawker](https://github.com/ahawker))
- Remove py34 support [\#415](https://github.com/ahawker/ulid/pull/415) ([ahawker](https://github.com/ahawker))
- Don't use tox with appveyor to fix py35 setup issue [\#414](https://github.com/ahawker/ulid/pull/414) ([ahawker](https://github.com/ahawker))
- Freeze pylint/pytest for py34 as its now deprecated [\#413](https://github.com/ahawker/ulid/pull/413) ([ahawker](https://github.com/ahawker))
- pyup.io:  Scheduled daily dependency update on Tuesday [\#412](https://github.com/ahawker/ulid/pull/412) ([pyup-bot](https://github.com/pyup-bot))

ulid/api/api.py
"""
    ulid/api/api
    ~~~~~~~~~~~~

    Contains functionality for public API methods for the 'ulid' package.
"""
import typing
import uuid

from .. import base32, codec, hints, providers, ulid

#: Type hint that defines multiple primitive types that can represent a full ULID.
ULIDPrimitive = typing.Union[hints.Primitive, uuid.UUID, ulid.ULID]  # pylint: disable=invalid-name

#: Defines the '__all__' for the API interface.
ALL = [
    'new',
    'parse',
    'create',
    'from_bytes',
    'from_int',
    'from_str',
    'from_uuid',
    'from_timestamp',
    'from_randomness',
    'MIN_TIMESTAMP',
    'MAX_TIMESTAMP',
    'MIN_RANDOMNESS',
    'MAX_RANDOMNESS',
    'MIN_ULID',
    'MAX_ULID',
    'Timestamp',
    'Randomness',
    'ULID'
]


class Api:
    """
    Encapsulates public API methods for the 'ulid' package that is agnostic to the underlying provider.
    """
    def __init__(self, provider: providers.Provider) -> None:
        """
        Create a new API instance with the given provider.

        :param provider: Provider that yields timestamp/randomness values.
        :type provider: :class:`~ulid.providers.Provider`
        """
        self.provider = provider

    def new(self) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance.

        The timestamp and randomness values are created from
        the instance :class:`~ulid.providers.Provider`.

        :return: ULID from current timestamp
        :rtype: :class:`~ulid.ulid.ULID`
        """
        timestamp = self.provider.timestamp()
        randomness = self.provider.randomness(timestamp)
        return ulid.ULID(timestamp + randomness)

    def parse(self, value: ULIDPrimitive) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance from the given value.

        .. note:: This method should only be used when the caller is trying to parse a ULID from
        a value when they're unsure what format/primitive type it will be given in.

        :param value: ULID value of any supported type
        :type value: :class:`~ulid.api.ULIDPrimitive`
        :return: ULID from value
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when unable to parse a ULID from the value
        """
        if isinstance(value, ulid.ULID):
            return value
        if isinstance(value, uuid.UUID):
            return self.from_uuid(value)
        if isinstance(value, str):
            len_value = len(value)
            if len_value == 36:
                return self.from_uuid(uuid.UUID(value))
            if len_value == 32:
                return self.from_uuid(uuid.UUID(value))
            if len_value == 26:
                return self.from_str(value)
            if len_value == 16:
                return self.from_randomness(value)
            if len_value == 10:
                return self.from_timestamp(value)
            raise ValueError('Cannot create ULID from string of length {}'.format(len_value))
        if isinstance(value, (int, float)):
            return self.from_int(int(value))
        if isinstance(value, (bytes, bytearray)):
            return self.from_bytes(value)
        if isinstance(value, memoryview):
            return self.from_bytes(value.tobytes())
        raise ValueError('Cannot create ULID from type {}'.format(value.__class__.__name__))

    def from_timestamp(self, timestamp: codec.TimestampPrimitive) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance using a timestamp value of a supported type.

        The following types are supported for timestamp values:

        * :class:`~datetime.datetime`
        * :class:`~int`
        * :class:`~float`
        * :class:`~str`
        * :class:`~memoryview`
        * :class:`~ulid.ulid.Timestamp`
        * :class:`~ulid.ulid.ULID`
        * :class:`~bytes`
        * :class:`~bytearray`

        :param timestamp: Unix timestamp in seconds
        :type timestamp: See docstring for types
        :return: ULID using given timestamp and new randomness
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when the value is an unsupported type
        :raises ValueError: when the value is a string and cannot be Base32 decoded
        :raises ValueError: when the value is or was converted to something 48 bits
        """
        timestamp = codec.decode_timestamp(timestamp)
        randomness = self.provider.randomness(timestamp.bytes)
        return self.create(timestamp, randomness)

    def from_randomness(self, randomness: codec.RandomnessPrimitive) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance using the given randomness value of a supported type.

        The following types are supported for randomness values:

        * :class:`~int`
        * :class:`~float`
        * :class:`~str`
        * :class:`~memoryview`
        * :class:`~ulid.ulid.Randomness`
        * :class:`~ulid.ulid.ULID`
        * :class:`~bytes`
        * :class:`~bytearray`

        :param randomness: Random bytes
        :type randomness: See docstring for types
        :return: ULID using new timestamp and given randomness
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when the value is an unsupported type
        :raises ValueError: when the value is a string and cannot be Base32 decoded
        :raises ValueError: when the value is or was converted to something 80 bits
        """
        timestamp = self.provider.timestamp()
        return self.create(timestamp, randomness)

    @staticmethod
    def create(timestamp: codec.TimestampPrimitive, randomness: codec.RandomnessPrimitive) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance using the given timestamp and randomness values.

        The following types are supported for timestamp values:

        * :class:`~datetime.datetime`
        * :class:`~int`
        * :class:`~float`
        * :class:`~str`
        * :class:`~memoryview`
        * :class:`~ulid.ulid.Timestamp`
        * :class:`~ulid.ulid.ULID`
        * :class:`~bytes`
        * :class:`~bytearray`

        The following types are supported for randomness values:

        * :class:`~int`
        * :class:`~float`
        * :class:`~str`
        * :class:`~memoryview`
        * :class:`~ulid.ulid.Randomness`
        * :class:`~ulid.ulid.ULID`
        * :class:`~bytes`
        * :class:`~bytearray`

        :param timestamp: Unix timestamp in seconds
        :type timestamp: See docstring for types
        :param randomness: Random bytes
        :type randomness: See docstring for types
        :return: ULID using given timestamp and randomness
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when a value is an unsupported type
        :raises ValueError: when a value is a string and cannot be Base32 decoded
        :raises ValueError: when a value is or was converted to incorrect bit length
        """
        timestamp = codec.decode_timestamp(timestamp)
        randomness = codec.decode_randomness(randomness)
        return ulid.ULID(timestamp.bytes + randomness.bytes)

    @staticmethod
    def from_bytes(value: hints.Buffer) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance from the given :class:`~bytes`,
        :class:`~bytearray`, or :class:`~memoryview` value.

        :param value: 16 bytes
        :type value: :class:`~bytes`, :class:`~bytearray`, or :class:`~memoryview`
        :return: ULID from buffer value
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when the value is not 16 bytes
        """
        length = len(value)
        if length != 16:
            raise ValueError('Expects bytes to be 128 bits; got {} bytes'.format(length))

        return ulid.ULID(value)

    @staticmethod
    def from_int(value: int) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance from the given :class:`~int` value.

        :param value: 128 bit integer
        :type value: :class:`~int`
        :return: ULID from integer value
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when the value is not a 128 bit integer
        """
        if value < 0:
            raise ValueError('Expects positive integer')

        length = (value.bit_length() + 7) // 8
        if length > 16:
            raise ValueError('Expects integer to be 128 bits; got {} bytes'.format(length))

        return ulid.ULID(value.to_bytes(16, byteorder='big'))

    @staticmethod
    def from_str(value: str) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance from the given :class:`~str` value.

        :param value: Base32 encoded string
        :type value: :class:`~str`
        :return: ULID from string value
        :rtype: :class:`~ulid.ulid.ULID`
        :raises ValueError: when the value is not 26 characters or malformed
        """
        return ulid.ULID(base32.decode_ulid(value))

    @staticmethod
    def from_uuid(value: uuid.UUID) -> ulid.ULID:
        """
        Create a new :class:`~ulid.ulid.ULID` instance from the given :class:`~uuid.UUID` value.

        :param value: UUIDv4 value
        :type value: :class:`~uuid.UUID`
        :return: ULID from UUID value
        :rtype: :class:`~ulid.ulid.ULID`
        """
        return ulid.ULID(value.bytes)
